[MM-25268] Implement prefix/suffix searching on teams and channel page in system console

hahmadia opened the pull request:
#### Summary
<!--
A description of what this pull request does.
-->
Implements prefix/suffix searching for team and channel page in the system console

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://mattermost.atlassian.net/browse/MM-25268

**Screenshots**
![image](https://user-images.githubusercontent.com/17804942/83289833-a3f23980-a1b3-11ea-9c9a-978d369d6401.png)

![image](https://user-images.githubusercontent.com/17804942/83289859-ae143800-a1b3-11ea-9940-41b0f9795311.png)

![image](https://user-images.githubusercontent.com/17804942/83289956-d7cd5f00-a1b3-11ea-8eff-ad1ebeb42bde.png)

sbishel commented on store/sqlstore/team_store.go:
```diff
@@ -364,7 +364,7 @@ func (s SqlTeamStore) SearchAll(term string) ([]*model.Team, *model.AppError) {
 	term = sanitizeSearchTerm(term, "\\")
 
-	if _, err := s.GetReplica().Select(&teams, "SELECT * FROM Teams WHERE Name LIKE :Term OR DisplayName LIKE :Term", map[string]interface{}{"Term": term + "%"}); err != nil {
+	if _, err := s.GetReplica().Select(&teams, "SELECT * FROM Teams WHERE Name LIKE :Term OR DisplayName LIKE :Term", map[string]interface{}{"Term": "%" + term + "%"}); err != nil {
```
I would build this in a small function. ie. wildcardSearchTerm(term)

hahmadia replied:
Good call. I have turned that into a function and added tests for both channel and teams.

mkraft commented:
@hahmadia Are these queries case insensitive? The queries don't look it and I wouldn't think functionally we want them to be case sensitive.

hahmadia commented:
> @hahmadia Are these queries case insensitive? The queries don't look it and I wouldn't think functionally we want them to be case sensitive.

I added a strings.ToLower(%term%) and added tests on both channels and teams to ensure case insensitive. Thanks for the point out!

hahmadia commented:
/update-branch

sbishel commented on api4/channel_test.go:
```diff
@@ -1150,11 +1150,53 @@ func TestSearchAllChannels(t *testing.T) {
+	assert.Len(t, *channels, 1)
+	assert.Equal(t, th.BasicChannel.Id, (*channels)[0].Id)
+
+	search = &model.ChannelSearch{Term: th.BasicChannel.Name[0:2] + strings.ToUpper(th.BasicChannel.Name[2:5]) + th.BasicChannel.Name[5:]}
```
You may want to throw in a comment here, just mentioning what you're testing. 
// Test Mixed Case.

sbishel commented on api4/channel_test.go:
```diff
@@ -1150,11 +1150,53 @@ func TestSearchAllChannels(t *testing.T) {
+
+	channels, resp = th.SystemAdminClient.SearchAllChannels(search)
+	CheckNoError(t, resp)
+
```
Add a couple more tests for thoroughness.
(ie, starts with, ends with, search for display name.)
 "what", "ever"

sbishel commented on api4/team_test.go:
```diff
@@ -1220,18 +1220,50 @@ func TestSearchAllTeamsPaged(t *testing.T) {
+			Search:             &model.TeamSearch{Term: "oob", Page: model.NewInt(0), PerPage: model.NewInt(100)},
+			ExpectedTeams:      []string{foobarTeam.Id},
+			ExpectedTotalCount: 1,
+		},
```
Same here, add search for "what" and "ever".

hahmadia commented:
> I would just a couple more tests for completeness.
> 
> Also, this may be handled in the web app. But do we only search after 2 or 3 characters are entered? Do we search on a single character?

We search whenever the user presses `Enter` on their keyboard.

hahmadia commented:
/update-branch

hahmadia commented:
/update-branch

hahmadia commented:
/branch-update

hahmadia commented:
/update-branch

catalintomai commented on store/sqlstore/team_store.go:
```diff
@@ -374,8 +378,14 @@ func (s SqlTeamStore) SearchAll(term string) ([]*model.Team, *model.AppError) {
+	term = wildcardSearchTerm(term)
 
-	if _, err := s.GetReplica().Select(&teams, "SELECT * FROM Teams WHERE Name LIKE :Term OR DisplayName LIKE :Term", map[string]interface{}{"Term": term + "%"}); err != nil {
+	operatorKeyword := "ILIKE"
```
I'd rather that we do not use Postgres (or any SQL DB provider) SQL extensions and stick as much as possible to vanilla SQL, otherwise porting the SQL code to any other provider (e.g. MSSQL) becomes an even more difficult task. Also, we should use/factor out `buildLIKEClause` and `buildFulltextClause` for both channels and teams

hahmadia replied:
Ticket created to keep track https://mattermost.atlassian.net/browse/MM-26727

catalintomai commented on api4/team_test.go:
```diff
@@ -1220,18 +1220,74 @@ func TestSearchAllTeamsPaged(t *testing.T) {
 		ExpectedTotalCount int64
 	}{
+		{
+			Name:               "Get foobar channel",
```
is it "team" instead of "channel"?

hahmadia replied:
Ticket made https://mattermost.atlassian.net/browse/MM-26720

agnivade commented on store/sqlstore/team_store.go:
```diff
@@ -374,8 +378,14 @@ func (s SqlTeamStore) SearchAll(term string) ([]*model.Team, *model.AppError) {
+	term = wildcardSearchTerm(term)
 
-	if _, err := s.GetReplica().Select(&teams, "SELECT * FROM Teams WHERE Name LIKE :Term OR DisplayName LIKE :Term", map[string]interface{}{"Term": term + "%"}); err != nil {
+	operatorKeyword := "ILIKE"
+	if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		operatorKeyword = "LIKE"
+	}
+	queryString := fmt.Sprintf("SELECT * FROM Teams WHERE Name %[1]s :Term OR DisplayName %[1]s :Term", operatorKeyword)
```
This pattern seems to be repeated throughout the PR. If possible, I'd prefer it to be refactored out to a separate place.

And also, construct the query using Squirrel if possible. Makes for much more readable SQL queries.

hahmadia replied:
Ticket created to keep track https://mattermost.atlassian.net/browse/MM-26726
